File: src/main/java/edu/kit/iti/formal/automation/testtables/model/GeneralizedTestTable.java
package edu.kit.iti.formal.automation.testtables.model;

/*-
 * #%L
 * geteta
 * %%
 * Copyright (C) 2016 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import edu.kit.iti.formal.automation.st.ast.FunctionDeclaration;
import edu.kit.iti.formal.automation.st.ast.TopLevelElement;
import edu.kit.iti.formal.automation.testtables.io.IOFacade;
import edu.kit.iti.formal.automation.testtables.model.options.PropertyInitializer;
import edu.kit.iti.formal.automation.testtables.model.options.TableOptions;
import edu.kit.iti.formal.automation.testtables.schema.ConstraintVariable;
import edu.kit.iti.formal.automation.testtables.schema.IoVariable;
import edu.kit.iti.formal.automation.testtables.schema.Variable;
import edu.kit.iti.formal.smv.ast.SMVExpr;
import edu.kit.iti.formal.smv.ast.SVariable;

import java.util.*;

/**
 * @author Alexander Weigl
 * @version 2
 */
public class GeneralizedTestTable {
    private final LinkedHashMap<String, IoVariable> ioVariables = new LinkedHashMap<>();
    private final Map<String, ConstraintVariable> constraintVariables = new HashMap<>();
    private final Map<String, SVariable> variableMap = new HashMap<>();
    private final Properties properties = new Properties(System.getProperties());
    private final Map<String, FunctionDeclaration> functions = new HashMap<>();
    private final Map<SVariable, Integer> references = new HashMap<>();
    private Region region;
    private TableOptions options = null;
    private String name;

    public TableOptions getOptions() {
        if(options==null) {
            options = new TableOptions();
            PropertyInitializer.initialize(options, properties);
        }
        return options;
    }

    public GeneralizedTestTable setOptions(TableOptions options) {
        this.options = options;
        return this;
    }

    public Map<String, IoVariable> getIoVariables() {
        return ioVariables;
    }

    public Map<String, ConstraintVariable> getConstraintVariable() {
        return constraintVariables;
    }

    public Region getRegion() {
        return region;
    }

    public void setRegion(Region region) {
        this.region = region;
    }

    public SVariable getSMVVariable(String text) {
        variableMap.computeIfAbsent(text,
                (k) -> IOFacade.asSMVVariable(getVariable(k)));
        return variableMap.get(text);
    }

    private Variable getVariable(String text) {
        IoVariable a = ioVariables.get(text);
        ConstraintVariable b = constraintVariables.get(text);

        if(a!=null && b!=null)
            throw new IllegalStateException("constraint and io variable have the same name.");

        if (a != null)
            return a;
        else
            return b;
    }

    public void add(IoVariable v) {
        ioVariables.put(v.getName(), v);
    }

    public void add(ConstraintVariable v) {
        constraintVariables.put(v.getName(), v);
    }

    public void addOption(String key, String value) {
        properties.put(key, value);
        options = null; // reset options
    }

    public void addFunctions(List<TopLevelElement> file) {
        for (TopLevelElement e : file) {
            functions.put(e.getBlockName(), (FunctionDeclaration) e);
        }
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public IoVariable getIoVariables(int i) {
        int k = 0;
        for (IoVariable v : ioVariables.values()) {
            if (k++ == i) return v;
        }
        return null;
    }

    public Map<SVariable, Integer> getReferences() {
        return references;
    }

    public SMVExpr getReference(SVariable columnVariable, int i) {
        if (i == 0) {
            return columnVariable;
        } else {
            SReference ref = new SReference(i, columnVariable);
            int max = Math.max(references.getOrDefault(columnVariable, i), i);
            references.put(columnVariable, max);
            return ref.asVariable();
        }
    }
}

File: src/main/kotlin/edu/kit/iti/formal/automation/testtables/model/GeneralizedTestTable.kt
/*
 * geteta
 *
 * Copyright (C) 2016-2018 -- Alexander Weigl <weigl@kit.edu>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 */
package edu.kit.iti.formal.automation.testtables.model


import edu.kit.iti.formal.automation.st.ast.FunctionDeclaration
import edu.kit.iti.formal.automation.st.ast.TopLevelElement
import edu.kit.iti.formal.automation.testtables.StateReachability
import edu.kit.iti.formal.automation.testtables.io.IOFacade
import edu.kit.iti.formal.automation.testtables.model.options.PropertyInitializer
import edu.kit.iti.formal.automation.testtables.model.options.TableOptions
import edu.kit.iti.formal.automation.testtables.schema.ConstraintVariable
import edu.kit.iti.formal.automation.testtables.schema.IoVariable
import edu.kit.iti.formal.automation.testtables.schema.Variable
import edu.kit.iti.formal.smv.ast.SMVExpr
import edu.kit.iti.formal.smv.ast.SVariable

import java.util.*

/**
 * @author Alexander Weigl
 * @version 2
 */
class GeneralizedTestTable {
    private val ioVariables = LinkedHashMap<String, IoVariable>()
    private val constraintVariables = HashMap<String, ConstraintVariable>()
    private val variableMap = HashMap<String, SVariable>()
    private val properties = Properties(
            System.getProperties())
    private val functions = HashMap<String, FunctionDeclaration>()
    private val references = HashMap<SVariable, Int>()
    var region: Region? = null
    var options: TableOptions? = null
        get() {
            if (field == null) {
                field = TableOptions()
                PropertyInitializer.initialize(field, properties)
            }
            return field
        }
    var name: String? = null

    val constraintVariable: Map<String, ConstraintVariable>
        get() = constraintVariables

    fun calculateReachability() {
        for (s in this.region!!.flat()) {
            s.outgoing.clear()
            s.incoming.clear()
            s.automataStates.clear()

            for (a in s.automataStates) {
                a.outgoing.clear()
                a.incoming.clear()
            }
        }
        StateReachability(this)
    }

    fun getIoVariables(): Map<String, IoVariable> {
        return ioVariables
    }

    fun getSMVVariable(text: String): SVariable {
        variableMap.computeIfAbsent(text) { k ->
            IOFacade.asSMVVariable(getVariable(k))
        }
        return variableMap[text]
    }

    private fun getVariable(text: String): Variable? {
        val a = ioVariables[text]
        val b = constraintVariables[text]

        if (a != null && b != null)
            throw IllegalStateException(
                    "constraint and io variable have the same name.")

        return a ?: b
    }

    fun add(v: IoVariable) {
        ioVariables[v.name] = v
    }

    fun add(v: ConstraintVariable) {
        constraintVariables[v.name] = v
    }

    fun addOption(key: String, value: String) {
        properties[key] = value
        options = null // reset options
    }

    fun addFunctions(file: List<TopLevelElement<*>>) {
        for (e in file) {
            functions[e.identifier] = e as FunctionDeclaration
        }
    }

    fun getIoVariables(i: Int): IoVariable? {
        var k = 0
        for (v in ioVariables.values) {
            if (k++ == i)
                return v
        }
        return null
    }

    fun getReferences(): Map<SVariable, Int> {
        return references
    }

    fun getReference(columnVariable: SVariable, i: Int): SMVExpr {
        if (i == 0) {
            return columnVariable
        } else {
            val ref = SReference(i, columnVariable)
            val max = Math.max((references as java.util.Map<SVariable, Int>).getOrDefault(columnVariable, i), i)
            references[columnVariable] = max
            return ref.asVariable()
        }
    }
}
